File: android/src/main/java/com/kakaotam/rnsdk/talk/TalkApiClientModule.kt
package com.kakaotam.rnsdk.talk

import com.facebook.react.bridge.*
import com.kakao.sdk.talk.TalkApiClient
import com.kakao.sdk.talk.model.*
import com.kakaotam.rnsdk.common.PromiseTransformer
import com.kakaotam.rnsdk.common.ReactToParam
import com.kakaotam.rnsdk.common.SdkInitializer
import com.kakaotam.rnsdk.talk.model.*

class TalkApiClientModule(private val reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    override fun getName(): String {
        return "TalkApiClientModule"
    }

    @ReactMethod
    private fun addChannelUrl(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, AddChannelParam::class.java)!!
        promise.resolve(TalkApiClient.instance.addChannelUrl(p.channelPublicId).toString())
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun channelChatUrl(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, ChannelChatParam::class.java)!!
        promise.resolve(TalkApiClient.instance.channelChatUrl(p.channelPublicId).toString())
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun channels(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, ChannelsParam::class.java)!!
        TalkApiClient.instance.channels(p.publicIds) f@{ relations: Channels?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(PromiseTransformer.toJsonString(relations))
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun friends(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, FriendsParam::class.java)!!
        TalkApiClient.instance.friends(
            p.offset,
            p.limit,
            p.order,
            p.friendOrder
        ) f@{ friends: Friends<Friend>?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(PromiseTransformer.toJsonString(friends))
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun profile(promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        TalkApiClient.instance.profile() f@{ profile: TalkProfile?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(PromiseTransformer.toJsonString(profile))
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun sendCustomMemo(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, SendCustomMemoParam::class.java)!!
        TalkApiClient.instance.sendCustomMemo(p.templateId, p.templateArgs) f@{ error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(true)
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun sendDefaultMemo(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, SendDefaultMemoParam::class.java)!!
        TalkApiClient.instance.sendDefaultMemo(p.templatable!!) f@{ error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(true)
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun sendScrapMemo(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, SendScrapMemoParam::class.java)!!
        TalkApiClient.instance.sendScrapMemo(p.requestUrl, p.templateId, p.templateArgs) f@{ error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(true)
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun sendCustomMessage(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, SendCustomMessageParam::class.java)!!
        TalkApiClient.instance.sendCustomMessage(
            p.receiverUuids,
            p.templateId,
            p.templateArgs
        ) f@{ messageSendResult: MessageSendResult?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(messageSendResult)
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun sendDefaultMessage(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, SendDefaultMessageParam::class.java)!!
        TalkApiClient.instance.sendDefaultMessage(
            p.receiverUuids,
            p.templatable!!,
        ) f@{ messageSendResult: MessageSendResult?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(messageSendResult)
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }

    @ReactMethod
    private fun sendScrapMessage(param: ReadableMap, promise: Promise) = try {
        SdkInitializer.ensure(reactContext)
        val p = ReactToParam.toParam(param, SendScrapMessageParam::class.java)!!
        TalkApiClient.instance.sendScrapMessage(
            p.receiverUuids,
            p.requestUrl,
            p.templateId,
            p.templateArgs
        ) f@{ messageSendResult: MessageSendResult?, error: Throwable? ->
            error?.let {
                promise.reject(PromiseTransformer.rnCode(it), PromiseTransformer.rnMessage(it), it)
                return@f
            }
            promise.resolve(messageSendResult)
        }
    } catch (e: Exception) {
        promise.reject(PromiseTransformer.rnCode(e), PromiseTransformer.rnMessage(e), e)
    }
}

File: ios/KakaoSdkTalk/TalkApiClientModule.swift
//
//  TalkApiClientModule.swift
//  react-native-kakao-sdk-all
//
//  Created by kakao-tam on 2022/05/06.
//

import Foundation

import KakaoSDKTalk

@objc(TalkApiClientModule)
class TalkApiClientModule: NSObject {

    @objc(addChannelUrl:withResolve:withReject:)
    func addChannelUrl(_ param:NSDictionary,
                       resolve:@escaping RCTPromiseResolveBlock,
                       reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(AddChannelParam.self)!
            if let data = TalkApi.shared.makeUrlForAddChannel(channelPublicId: p.channelPublicId) {
                resolve(data.absoluteString)
            } else {
                resolve("")
            }
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(channelChatUrl:withResolve:withReject:)
    func channelChatUrl(_ param:NSDictionary,
                        resolve:@escaping RCTPromiseResolveBlock,
                        reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(ChannelChatParam.self)!
            if let data = TalkApi.shared.makeUrlForChannelChat(channelPublicId: p.channelPublicId) {
                resolve(data.absoluteString)
            } else {
                resolve("")
            }
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(channels:withResolve:withReject:)
    func channels(_ param:NSDictionary,
                  resolve:@escaping RCTPromiseResolveBlock,
                  reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(ChannelsParam.self)!
            TalkApi.shared.channels(publicIds: p.publicIds) {
                (channels: Channels?, error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(channels?.toJsonString());
                }
            }
            
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(friends:withResolve:withReject:)
    func friends(_ param:NSDictionary,
                 resolve:@escaping RCTPromiseResolveBlock,
                 reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(FriendsParam.self)!
            TalkApi.shared.friends(offset: p.offset,
                                   limit: p.limit,
                                   order: p.order,
                                   friendOrder: p.friendOrder) {
                (friends: Friends<Friend>?, error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(friends?.toJsonString());
                }
            }
            
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(profile:withReject:)
    func profile(_ resolve:@escaping RCTPromiseResolveBlock,
                 reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            TalkApi.shared.profile() {
                (talkProfile: TalkProfile?, error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(talkProfile?.toJsonString());
                }
            }
            
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(sendCustomMemo:withResolve:withReject:)
    func sendCustomMemo(_ param:NSDictionary,
                        resolve:@escaping RCTPromiseResolveBlock,
                        reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(SendCustomMemoParam.self)!
            TalkApi.shared.sendCustomMemo(templateId: p.templateId,
                                          templateArgs: p.templateArgs) {
                (error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(true);
                }
            }
            
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(sendDefaultMemo:withResolve:withReject:)
    func sendDefaultMemo(_ param:NSDictionary,
                         resolve:@escaping RCTPromiseResolveBlock,
                         reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(SendDefaultMemoParam.self)!
            TalkApi.shared.sendDefaultMemo(templatable: p.templatable) {
                (error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(true);
                }
            }
            
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(sendScrapMemo:withResolve:withReject:)
    func sendScrapMemo(_ param:NSDictionary,
                       resolve:@escaping RCTPromiseResolveBlock,
                       reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(SendScrapMemoParam.self)!
            TalkApi.shared.sendScrapMemo(requestUrl: p.requestUrl,
                                         templateId: p.templateId,
                                         templateArgs: p.templateArgs) {
                (error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(true);
                }
            }
            
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(sendCustomMessage:withResolve:withReject:)
    func sendCustomMessage(_ param:NSDictionary,
                           resolve:@escaping RCTPromiseResolveBlock,
                           reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(SendCustomMessageParam.self)!
            TalkApi.shared.sendCustomMessage(templateId: p.templateId,
                                             templateArgs: p.templateArgs,
                                             receiverUuids: p.receiverUuids) {
                (messageSendResult: MessageSendResult?, error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(messageSendResult?.toJsonString());
                }
            }
            
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(sendDefaultMessage:withResolve:withReject:)
    func sendDefaultMessage(_ param:NSDictionary,
                            resolve:@escaping RCTPromiseResolveBlock,
                            reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(SendDefaultMessageParam.self)!
            TalkApi.shared.sendDefaultMessage(templatable: p.templatable,
                                              receiverUuids: p.receiverUuids) {
                (messageSendResult: MessageSendResult?, error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(messageSendResult?.toJsonString());
                }
            }
            
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
    
    @objc(sendScrapMessage:withResolve:withReject:)
    func sendScrapMessage(_ param:NSDictionary,
                          resolve:@escaping RCTPromiseResolveBlock,
                          reject:@escaping RCTPromiseRejectBlock) -> Void {
        let cErr = ObjC.tryCatch {
            SdkInitializer.shared.ensure()
            let p = param.toParam(SendScrapMessageParam.self)!
            TalkApi.shared.sendScrapMessage(requestUrl: p.requestUrl,
                                            templateId: p.templateId,
                                            templateArgs: p.templateArgs,
                                            receiverUuids: p.receiverUuids) {
                (messageSendResult: MessageSendResult?, error: Error?) in
                if let error = error {
                    reject(error.rnCode(), error.rnMessage(), error)
                } else {
                    resolve(messageSendResult?.toJsonString());
                }
            }
            
        }
        if let cErr = cErr {
            reject(cErr.rnCode(), cErr.rnMessage(), cErr)
        }
    }
}
